wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala

```diff
@@ -253,21 +265,36 @@ private[wallet] trait TransactionProcessing extends WalletLogger {
   /** If the given UTXO is marked as unspent, updates
     * its spending status. Otherwise returns `None`.
     */
-  private def markAsPendingSpent(
-      out: SpendingInfoDb): Future[Option[SpendingInfoDb]] = {
+  private def markAsSpent(
+      out: SpendingInfoDb,
+      spendingTxId: DoubleSha256DigestBE): Future[Option[SpendingInfoDb]] = {
     out.state match {
       case TxoState.ConfirmedReceived | TxoState.PendingConfirmationsReceived |
           TxoState.ImmatureCoinbase =>
         val updated =
-          out.copyWithState(state = TxoState.PendingConfirmationsSpent)
+          out
+            .copyWithState(state = TxoState.PendingConfirmationsSpent)
+            .copyWithSpendingTxId(spendingTxId)
         val updatedF =
           spendingInfoDAO.update(updated)
         updatedF.foreach(updated =>
           logger.debug(
             s"Marked utxo=${updated.toHumanReadableString} as state=${updated.state}"))
         updatedF.map(Some(_))
-      case TxoState.Reserved | TxoState.ConfirmedSpent |
-          TxoState.PendingConfirmationsSpent | TxoState.DoesNotExist =>
+      case TxoState.Reserved | TxoState.PendingConfirmationsSpent =>
+        val updated =
+          out.copyWithSpendingTxId(spendingTxId)
+        val updatedF =
+          spendingInfoDAO.update(updated)
+        updatedF.map(Some(_))
+      case TxoState.DoesNotExist | TxoState.ConfirmedSpent =>
+        if (
+          out.spendingTxIdOpt.isDefined && !out.spendingTxIdOpt.contains(
+            spendingTxId)
+        ) {
+          logger.warn(
+            s"Attempted to mark an already spent utxo ${out.outPoint.hex} with a new spending tx ${spendingTxId.hex}")
```

Comment:
Let's make unique log messages for both `TxoState.DoesNotExist` and `TxoState.ConfirmedSpent` since they are pretty different situations. 

it seems we should also be recommending rescans of the wallet at minimum if these cases occur? 

This is getting into wallet corruption territory, maybe we want to return an exception rather than `FutureUtil.none`?